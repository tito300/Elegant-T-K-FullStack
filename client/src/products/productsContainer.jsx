import React, { Component } from 'react';
import { connect } from 'react-redux';
import axios from 'axios';
import ProductCard from './productCard';
import propTypes from 'prop-types';

export class Products extends Component {
    state = {
        products: [],
        allProducts: [],
        page: 1,
        numberOfPages: 4,
        itemsPerPage: 6,
        category: "top-sellers",
        error: false,
        errMsg: "",
        pending: false,
    }

    static propTypes = {
        data: propTypes.array,
    }

    async componentDidMount() {
        this.setState({ pending: true });
        let products = await this.getProducts();
        let numberOfPages = Math.ceil(products.length / this.state.itemsPerPage);
        this.setState({ allProducts: products, numberOfPages }, () => {
            this.updateCurrentPageProducts()
        })
    }

    componentDidUpdate(prevprops, prevstate) {
        if (prevstate.page !== this.state.page) {
            this.updateCurrentPageProducts();
        }
    }

    handlePageChange = (e) => {
        e.preventDefault();
        let id = e.target.id === "" ? e.target.parentNode.id : e.target.id;
        this.setState({ page: parseInt(id) });
    }

    getProducts = async () => {
        const url = `/api/products/${this.state.category}?page=${this.state.page}&productsPerReq=${this.state.itemsPerPage * this.state.numberOfPages}`

        let res = await axios.get(url)
        let data = [...res.data];
        let products = [...this.state.allProducts, ...data];

        // TODO: implement friendly error using Error boundaries comp
        if (data.length === undefined) return this.setState({ error: true, errMsg: "server response was not proper, try to refresh the page" });

        // NOTSURE: whether there is a need to keep products in redux;
        // this.props.populateProducts(products);

        return products;

    }

    updateCurrentPageProducts = () => {
        let { page, itemsPerPage, allProducts } = this.state;
        let all = [...allProducts]
        let result = all.slice(((page - 1) * itemsPerPage), (itemsPerPage * page))

        this.setState({ products: result, allProducts: all, pending: false });
        return result;
    }


    render() {
        let { page, products, pending, numberOfPages } = this.state;

        let getPagingElements = () => {
            let result = [];
            for (let i = 1; i <= numberOfPages; i++) {
                let class1 = page === i ? "pager-li active" : "pager-li"
                let num = i.toString();
                result.push(<li id={num} key={num} className={class1} onClick={this.handlePageChange}>
                    <a href="#" >{num}</a>
                </li>)
            }
            return result;
        }

        return (
            <div className="body-section">
                <h1 className="page-title">TOP SELLERS</h1>
                <div className="products">
                    {products.length > 0 ? products.map((item, i) => { // makes sure data exists
                        return (
                            <ProductCard
                                key={i}
                                i={i}
                                item={item}
                            />
                        )
                    })
                        : (products.length === 0) && !pending ? <p className='text-error'>Sorry, there is no more products to show</p>
                            : <p className="fetching-items">Fetching Items...</p>}
                </div>

                {/* TODO: extract this pager to a pure component to make it reusable */}
                <div className="pager-container">
                    <ul className="pager-ul">
                        {getPagingElements()}
                    </ul>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, ownProps) => {
    return {
        data: state.products.all,
        loggedin: state.active,
    };
}

const mapDispatchToProps = (dispatch) => {
    return {
        populateProducts: (products) => { dispatch({ type: 'POPULATE_PRODUCTS', products: products }) }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Products);
