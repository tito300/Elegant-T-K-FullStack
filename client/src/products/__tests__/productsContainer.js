import { shallow } from 'enzyme';
import React from 'react';
import { Products } from '../productsContainer';
let wrapper;

describe('ProductsContainer component', ()=>{
    beforeAll(()=>{
        wrapper = shallow(<Products />, {
            disableLifecycleMethods: true
        });
    })

    it('should render ProductsContainer', () => {
        expect(wrapper).toMatchSnapshot();  
    })

    it('initial default state should be correct', ()=>{
        expect.assertions(9)

        expect(wrapper.state('page')).toBe(1);
        expect(wrapper.state('products').length).toBe(0);
        expect(wrapper.state('allProducts').length).toBe(0);
        expect(wrapper.state('numberOfPages')).toBeDefined();
        expect(wrapper.state('itemsPerPage')).toBeDefined();
        expect(wrapper.state('category')).toBeDefined();
        expect(wrapper.state('error')).toBeFalsy();
        expect(wrapper.state('errMsg')).toBe("");
        expect(wrapper.state('pending')).toBeFalsy();
        // console.log(wrapper.instance());
    })

    it('should show "no items available"', ()=> {
        expect(wrapper.find('.text-error').exists()).toBeTruthy();
    } )
    
    it('should show fetching', () => {
        expect.assertions(1);

        wrapper.setState({ pending: true });
        expect(wrapper.find('.fetching-items').exists()).toBeTruthy();
    })

    it('should show products when recieved', async (done) => {
        expect.assertions(1);
        
        wrapper.instance().getProducts = jest.fn().mockResolvedValue(products);
        wrapper.setProps({
            match: {
                params: { category: 'topsellers' }
            }
        })
        await wrapper.instance().componentDidMount();
        // expect(wrapper).toMatchSnapshot('temp snapshot');

        expect(wrapper.find('.products').children().length).toBe(6);
        done();
    })

    it('should update page count and class when page 2 is clicked', ()=>{
        expect.assertions(1);
        
        wrapper.instance().handlePageChange({
            target: { id: '2' },
        })
        expect(wrapper.state('page')).toBe(2);
        // expect(wrapper.find('li[id="2"]').hasClass('active')).toBeTruthy();
    })
})


var products = [
    {
        title: "random",
        id: '23442',
        discreption: 'sdsdsdsd',
        photo: './img/ssdsdq.jpg'
    },
    {
        title: "random",
        id: '23442',
        discreption: 'sdsdsdsd',
        photo: './img/ssdsdq.jpg'
    },
    {
        title: "random",
        id: '23442',
        discreption: 'sdsdsdsd',
        photo: './img/ssdsdq.jpg'
    },
    {
        title: "random",
        id: '23442',
        discreption: 'sdsdsdsd',
        photo: './img/ssdsdq.jpg'
    },
    {
        title: "random",
        id: '23442',
        discreption: 'sdsdsdsd',
        photo: './img/ssdsdq.jpg'
    },
    {
        title: "random",
        id: '23442',
        discreption: 'sdsdsdsd',
        photo: './img/ssdsdq.jpg'
    },
    {
        title: "random",
        id: '23442',
        discreption: 'sdsdsdsd',
        photo: './img/ssdsdq.jpg'
    },
]