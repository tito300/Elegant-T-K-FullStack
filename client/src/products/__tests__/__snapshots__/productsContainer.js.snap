// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsContainer component should render ProductsContainer 1`] = `
<div
  className="body-section"
>
  <h1
    className="page-title"
  >
    TOP SELLERS
  </h1>
  <div
    className="products"
  >
    <p
      className="text-error"
    >
      Sorry, there is no more products to show
    </p>
  </div>
  <div
    className="pager-container"
  >
    <ul
      className="pager-ul"
    >
      <li
        className="pager-li active"
        id="1"
        key="1"
        onClick={[Function]}
      >
        <a
          href="#"
        >
          1
        </a>
      </li>
      <li
        className="pager-li"
        id="2"
        key="2"
        onClick={[Function]}
      >
        <a
          href="#"
        >
          2
        </a>
      </li>
      <li
        className="pager-li"
        id="3"
        key="3"
        onClick={[Function]}
      >
        <a
          href="#"
        >
          3
        </a>
      </li>
      <li
        className="pager-li"
        id="4"
        key="4"
        onClick={[Function]}
      >
        <a
          href="#"
        >
          4
        </a>
      </li>
    </ul>
  </div>
</div>
`;
